### PR TITLE
Re-add overload for fixed-size arrays to `std.string.indexOf`

### PR DESCRIPTION
**Pull request by schuetzm**
Fixed the regression introduced in https://github.com/D-Programming-Language/phobos/pull/3172

### CODE REVIEW

**Comment by schveiguy** [comment]
Hm.. I think this is not what was asked for. The original code did not instantiate a new template per static array size. This is kind of a very tricky issue, because you want to use `T[]` not `T[N]`, but `T[]` is a valid range, so it will match the first overload also.

The only good news here is that the template bloat should be minimal in terms of code size (the function should always inline), but it will hurt in the symbol table. If ever there was a case for `@forceinline`, this is it.

I'm OK with this change, the template bloat should not be that bad. If others feel differently, I'd love to hear another solution.

**Comment by schuetzm** [comment]
I could factor the actual implementation out into a third, private template that is called by both of the others. But IMO it's not worth the trouble, and it would even make the common case worse, because it would instantiate two templates for each type.

File: std/string.d [hunk 1]
```diff
@@ -466,6 +466,13 @@ ptrdiff_t indexOf(Range)(Range s, in dchar c,
     return -1;
 }
 
+ptrdiff_t indexOf(T, size_t n)(T[n] s, in dchar c,
```
Comment (WalterBright) [inline std/string.d]: use:
```
ref T[n] s
```
instead, otherwise the entire array gets copied to the stack!

Comment (schveiguy) [inline std/string.d]: Oh right, good catch!

**Comment by schveiguy** [comment]
> But IMO it's not worth the trouble

yeah, agreed.

**Comment by schveiguy** [comment]
LGTM with latest changes.

**Comment by JakobOvrum** [comment]
You'd think that having an overload receiving `T[]` would work properly as it is more specialized than just `Range`, but DMD doesn't implement it...

**Comment by JinShil** [comment]
This pull request introduced a regression:  https://issues.dlang.org/show_bug.cgi?id=14735

**Comment by schuetzm** [comment]
Can someone figure this out? Removing the overload added in this PR makes it work, but this overload shouldn't match for the test case in the bug report, so I don't see how it could even have an effect. Is this a compiler bug?

**Comment by JinShil** [comment]
I don't fully understand what's going on here either, but according to my tests, if you add a default value to `startIdx` (i.e. `startIdx = 0`) at https://github.com/schuetzm/phobos/blob/97ddddeaebe3c095982d367164d9f6e933bb4b44/std/string.d#L537 It should compile.

Example:
```D
// Uncomment to resolve compiler error
/* ptrdiff_t indexOf(Char)(const(Char)[] s, in dchar c, in size_t startIdx = 0) @safe pure
{
    return 0;
} */

//  Error: template example.indexOf cannot deduce function from argument types !()(char[], char)
ptrdiff_t indexOf(T, size_t n)(ref T[] s, in char cs)
{
    return 0;
}

void main()
{
  char[64] foo;
  long i = foo[].indexOf('a');
}
```

**Comment by JinShil** [comment]
The problem has something to do with the `CaseSensitive` argument.  See the bug report for a test case.

**Comment by MartinNowak** [comment]
Was this the only regression regarding static arrays of newly rangified functions?

File: std/string.d [hunk 2]
```diff
@@ -466,6 +466,13 @@ ptrdiff_t indexOf(Range)(Range s, in dchar c,
     return -1;
 }
 
+ptrdiff_t indexOf(T, size_t n)(ref T[n] s, in dchar c,
+        in CaseSensitive cs = CaseSensitive.yes) @safe pure
+    if (isSomeChar!T)
+{
+    return indexOf(s[], c, cs);
```
Comment (MartinNowak) [inline std/string.d]: An optimization would be to use `const(T)[]`, b/c of less template instantiations.

**Comment by MartinNowak** [comment]
I think a better alternative would be to add explicit function overloads for `const(char/wchar/dchar)[]`.
That makes sure the overload is more specialized, we can precompile the most common instantiations into libphobos, and it allows to coerce to `const(char)[]` to deduplicate all const variations of narrow strings.

```d
void indexOf(Range)(Range s, in dchar c) @safe pure
{
    pragma(msg, "Range "~Range.stringof);
}

void indexOf(const(char)[] s, in dchar c) @safe pure
{
    pragma(msg, "const(char)[]");
    .indexOf!(const(char)[])(s, c);
}

unittest
{
    char[] ary;
    char[128] sary;
    indexOf(ary);
    indexOf(sary);
}
```

**Comment by MartinNowak** [comment]
This is troublesome b/c of polysemeous string literals and conversion of other arguments.

```d
void test(const(char)[] s, in dchar c) {}
void test(const(dchar)[] s, in dchar c) {}

void test()
{
    test("foo", 'a'); // 2nd arguments matches with conversion, so both overloads match equally well
}
```

**Pull request merged**